curvefs/src/metaserver/inode_manager.cpp

```diff
@@ -273,13 +266,71 @@ void RemoveFromS3ChunkInfoList(const S3ChunkInfoList &listToRemove,
     listToMerge->Swap(&newList);
 }
 
+int ProcessRequestFromS3Compact(
+    const google::protobuf::Map<uint64_t, S3ChunkInfoList>& s3ChunkInfoAdd,
+    const google::protobuf::Map<uint64_t, S3ChunkInfoList>& s3ChunkInfoRemove,
+    Inode* inode) {
+    VLOG(9) << "s3compact: request from s3compaction, inodeId:"
+            << inode->inodeid();
+    auto checkWithLog = [&](bool cond) {
+        if (!cond)
+            LOG(ERROR) << "s3compact: bad GetOrModifyS3ChunkInfo request";
+        return cond;
+    };
+    if (!checkWithLog(s3ChunkInfoAdd.size() == s3ChunkInfoRemove.size()))
+        return -1;
+    for (const auto& item : s3ChunkInfoAdd) {
+        const auto& chunkIndex = item.first;
+        const auto& toAddList = item.second;
+        if (!checkWithLog(toAddList.s3chunks_size() == 1)) return -1;
+        if (!checkWithLog(s3ChunkInfoRemove.find(chunkIndex) !=
+                          s3ChunkInfoRemove.end()))
+            return -1;
+        const auto& toRemoveList = s3ChunkInfoRemove.at(chunkIndex);
+        const auto& origItem = inode->s3chunkinfomap().find(chunkIndex);
+        if (!checkWithLog(origItem != inode->s3chunkinfomap().end())) return -1;
+        // delete and insert
+        const auto& compactChunkId = toAddList.s3chunks(0).chunkid();
+        S3ChunkInfoList newList;
+        const auto& origList = origItem->second;
+        bool inserted = false;
+        for (int i = 0; i < origList.s3chunks_size(); i++) {
+            const auto& origS3chunk = origList.s3chunks(i);
+            const auto& toRemoveS3chunks = toRemoveList.s3chunks();
+            auto s3chunkIt =
+                std::find_if(toRemoveS3chunks.begin(), toRemoveS3chunks.end(),
+                             [&](const S3ChunkInfo& toRemoveS3chunk) {
+                                 return MessageDifferencer::Equals(
+                                     toRemoveS3chunk, origS3chunk);
+                             });
+            if (s3chunkIt == toRemoveS3chunks.end()) {
+                if (origS3chunk.chunkid() > compactChunkId && !inserted) {
+                    *newList.add_s3chunks() = toAddList.s3chunks(0);
+                    VLOG(9) << "s3compact: add chunkid "
+                            << toAddList.s3chunks(0).chunkid();
+                    inserted = true;
```

Comment:
if the compacted s3chunkinfo is inserted, does it means all behind s3chunkinfos in origList is newer ? if so, you can just add all newer s3chunkinfos into newList once.

Reply:
yes, add a fastpath for the left ones